### PR TITLE
CB-16136:Add cloud provider type to cm settings

### PR DESCRIPTION
**Pull request by satyabollineni**
Added the -cp option to indicate the cloud provider to CM server file.

The screenshot shared below is from the file /etc/default/cloudera-scm-server file
Before the change:
image

After the change :

`CMF_SERVER_ARGS="-i /etc/cloudera-scm-server/cm.settings -cp  AWS -env PUBLIC_CLOUD"`

### CODE REVIEW

**Review by horadla23 — APPROVED** [review]
lgtm

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java [hunk 1]
```diff
@@ -656,12 +659,22 @@ public Map<String, SaltPillarProperties> createPillarWithClouderaManagerSettings
                                 "missed_heartbeat_interval", cmMissedHeartbeatInterval,
                                 "disable_auto_bundle_collection", disableAutoBundleCollection,
                                 "set_cdp_env", isVersionNewerOrEqualThanLimited(cmVersion, CLOUDERAMANAGER_VERSION_7_0_2),
+                                "cloud_provider_setup_supported", isCloudProviderSetupSupported(stackDto, cmVersion),
+                                "cloud_provider", stackDto.getCloudPlatform(),
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:663]: Cloud provider (e.g. the platform) is already available at the top level pillar, with `salt['pillar.get']('platform')`

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java [hunk 2]
```diff
@@ -656,12 +659,22 @@ public Map<String, SaltPillarProperties> createPillarWithClouderaManagerSettings
...
                                 "deterministic_uid_gid", isVersionNewerOrEqualThanLimited(cmVersion, CLOUDERAMANAGER_VERSION_7_2_1),
                                 "cloudera_scm_sudo_access", CMRepositoryVersionUtil.isSudoAccessNeededForHostCertRotation(clouderaManagerRepo)),
                         "mgmt_service_directories", serviceLocations.getAllVolumePath(),
                         "address", primaryGatewayConfig.getPrivateAddress()))));
     }
 
+    private boolean isCloudProviderSetupSupported(StackDto stackDto, String cmVersion) {
+        final List<String> supportedCloudProviders = Arrays.asList(CloudPlatform.AWS.name(), CloudPlatform.GCP.name(), CloudPlatform.AZURE.name());
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:671]: No `final` please. Use `List.of(...)` if needed.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java [hunk 3]
```diff
@@ -656,12 +659,22 @@ public Map<String, SaltPillarProperties> createPillarWithClouderaManagerSettings
...
+        if (supportedCloudProviders.contains(stackDto.getCloudPlatform()) && isVersionNewerOrEqualThanLimited(cmVersion, CLOUDERAMANAGER_VERSION_7_6_2)) {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:672]: Simplify please. If the return value depends on the condition then you can return the evaluated condition.

File: core/src/test/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunnerTest.java [hunk 4]
```diff
@@ -319,6 +321,62 @@ void testDecoratePillarWithClouderaManagerRepoWithNoJsonLicense() throws IOExcep
...
+        Boolean cloudProviderSetupSupported = extractCloudPlatformSupported(clouderaManagerSettings);
+        assertEquals(cloudProviderSetupSupported, Boolean.FALSE);
+    }
+
+    private Boolean extractCloudPlatformSupported(Map<String, SaltPillarProperties> clouderaManagerSettings) {
+        SaltPillarProperties saltPillarProperties = clouderaManagerSettings.get("cloudera-manager-settings");
+        Map<String, Object> clouderaManager = (Map<String, Object>) saltPillarProperties.getProperties().get("cloudera-manager");
+        Map<Object, Object> settings = (Map<Object, Object>) clouderaManager.get("settings");
+        boolean cloudProviderSetupSupported = (boolean) settings.get("cloud_provider_setup_supported");
+        return  cloudProviderSetupSupported;
```
Comment (bergerdenes) [inline core/src/test/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunnerTest.java:377]: multiple spaces

File: core/src/test/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunnerTest.java [hunk 5]
```diff
@@ -319,6 +321,62 @@ void testDecoratePillarWithClouderaManagerRepoWithNoJsonLicense() throws IOExcep
...
+        Map<String, SaltPillarProperties> clouderaManagerSettings = underTest.createPillarWithClouderaManagerSettings(clouderaManagerRepo, stack, gatewayConfig);
+        Boolean cloudProviderSetupSupported = extractCloudPlatformSupported(clouderaManagerSettings);
+        assertEquals(cloudProviderSetupSupported, Boolean.FALSE);
+    }
+
+    private Boolean extractCloudPlatformSupported(Map<String, SaltPillarProperties> clouderaManagerSettings) {
+        SaltPillarProperties saltPillarProperties = clouderaManagerSettings.get("cloudera-manager-settings");
+        Map<String, Object> clouderaManager = (Map<String, Object>) saltPillarProperties.getProperties().get("cloudera-manager");
+        Map<Object, Object> settings = (Map<Object, Object>) clouderaManager.get("settings");
+        boolean cloudProviderSetupSupported = (boolean) settings.get("cloud_provider_setup_supported");
```
Comment (bergerdenes) [inline core/src/test/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunnerTest.java:376]: Why is the function return type a Boolean if you are returning a primitive?

File: orchestrator-salt/src/main/resources/salt/salt/cloudera/manager/init.sls [hunk 6]
```diff
@@ -34,7 +34,10 @@ add_settings_file_to_cfm_server_args:
   file.replace:
     - name: /etc/default/cloudera-scm-server
     - pattern: "CMF_SERVER_ARGS=.*"
-{% if salt['pillar.get']('cloudera-manager:settings:set_cdp_env') == True %}
+{% if salt['pillar.get']('cloudera-manager:settings:cloud_provider_setup_supported') == True %}
+    - repl: CMF_SERVER_ARGS="-i /etc/cloudera-scm-server/cm.settings -cp {{ cloudera_manager.settings.cloud_provider }} -env PUBLIC_CLOUD"
+    - unless: grep "CMF_SERVER_ARGS=\"-i /etc/cloudera-scm-server/cm.settings -cp {{ cloudera_manager.settings.cloud_provider }}\"" /etc/default/cloudera-scm-server
+{% elif salt['pillar.get']('cloudera-manager:settings:set_cdp_env') == True and salt['pillar.get']('cloudera-manager:settings:cloud_provider_setup_supported') == False %}
     - repl: CMF_SERVER_ARGS="-i /etc/cloudera-scm-server/cm.settings -env PUBLIC_CLOUD"
```
Comment (bergerdenes) [inline orchestrator-salt/src/main/resources/salt/salt/cloudera/manager/init.sls:41]: I feel that the whole logic is faulty here. It would set the args even if the set_cdp_env is false...

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java (hunk 2 above)
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java]: List.of is still not used

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java (hunk 3 above)
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java]: The outmost parentheses are not required.

File: core/src/test/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunnerTest.java [hunk 7]
```diff
@@ -319,6 +321,62 @@ void testDecoratePillarWithClouderaManagerRepoWithNoJsonLicense() throws IOExcep
...
+        when(clusterApiConnectors.getConnector(cluster)).thenReturn(clusterPreCreationApi);
+        when(clusterPreCreationApi.getServiceLocations()).thenReturn(serviceLocationMap);
+        when(serviceLocationMap.getAllVolumePath()).thenReturn(serviceLocations);
+        when(gatewayConfig.getPrivateAddress()).thenReturn("privateAddress");
+        Map<String, SaltPillarProperties> clouderaManagerSettings = underTest.createPillarWithClouderaManagerSettings(clouderaManagerRepo, stack, gatewayConfig);
+        Object cloudProviderSetupSupported = extractCloudPlatformSupported(clouderaManagerSettings);
+        assertEquals(cloudProviderSetupSupported, Boolean.FALSE);
+    }
+
+    private Object extractCloudPlatformSupported(Map<String, SaltPillarProperties>clouderaManagerSettings) {
```
Comment (bergerdenes) [inline core/src/test/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunnerTest.java:372]: why did you change to Object? It was a Boolean before.
It should be a primitive `boolean`

**Review by bergerdenes — CHANGES_REQUESTED** [review]

**Comment by satyabollineni** [comment]
retest

**Comment by satyabollineni** [comment]
retest this please

**Comment by satyabollineni** [comment]
retest this please

**Comment by satyabollineni** [comment]
retest this please

**Review by bergerdenes — APPROVED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java [hunk 8]
```diff
@@ -662,12 +664,18 @@ public Map<String, SaltPillarProperties> createPillarWithClouderaManagerSettings
...
                                 "set_cdp_env", isVersionNewerOrEqualThanLimited(cmVersion, CLOUDERAMANAGER_VERSION_7_0_2),
+                                "cloud_provider_setup_supported", isCloudProviderSetupSupported(stackDto, cmVersion),
                                 "deterministic_uid_gid", isVersionNewerOrEqualThanLimited(cmVersion, CLOUDERAMANAGER_VERSION_7_2_1),
                                 "cloudera_scm_sudo_access", CMRepositoryVersionUtil.isSudoAccessNeededForHostCertRotation(clouderaManagerRepo)),
                         "mgmt_service_directories", serviceLocations.getAllVolumePath(),
                         "address", primaryGatewayConfig.getPrivateAddress()))));
     }
 
+    private boolean isCloudProviderSetupSupported(StackDto stackDto, String cmVersion) {
+        List<String> supportedCloudProviders = List.of(CloudPlatform.AWS.name(), CloudPlatform.GCP.name(), CloudPlatform.AZURE.name());
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:675]: this should be a constant

File: orchestrator-salt/src/main/resources/salt/salt/cloudera/manager/init.sls [hunk 9]
```diff
@@ -44,7 +44,10 @@ add_settings_file_to_cfm_server_args:
   file.replace:
     - name: /etc/default/cloudera-scm-server
     - pattern: "CMF_SERVER_ARGS=.*"
-{% if salt['pillar.get']('cloudera-manager:settings:set_cdp_env') == True %}
+{% if salt['pillar.get']('cloudera-manager:settings:cloud_provider_setup_supported') == True %}
+    - repl: CMF_SERVER_ARGS="-i /etc/cloudera-scm-server/cm.settings -cp {{ salt['pillar.get']('platform') }} -env PUBLIC_CLOUD"
+    - unless: grep "CMF_SERVER_ARGS=\"-i /etc/cloudera-scm-server/cm.settings -cp {{ salt['pillar.get']('platform') }}\"" /etc/default/cloudera-scm-server
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply orchestrator-salt/src/main/resources/salt/salt/cloudera/manager/init.sls:49]: but as I see this was the case until now, so it shouldn't matter

File: orchestrator-salt/src/main/resources/salt/salt/cloudera/manager/init.sls [hunk 10]
```diff
@@ -44,7 +44,10 @@ add_settings_file_to_cfm_server_args:
...
+{% elif salt['pillar.get']('cloudera-manager:settings:set_cdp_env') == True and salt['pillar.get']('cloudera-manager:settings:cloud_provider_setup_supported') == False %}
```
Comment (lacikaaa) [inline orchestrator-salt/src/main/resources/salt/salt/cloudera/manager/init.sls:50]: `and salt['pillar.get']('cloudera-manager:settings:cloud_provider_setup_supported') == False` this part is not required

Comment (satyabollineni) [reply orchestrator-salt/src/main/resources/salt/salt/cloudera/manager/init.sls]: Addressed, plz review

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java (hunk 8 above)
Comment (lacikaaa) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:675]: what kind of failures did you experience? I'm not able to see how this could cause any issue

Comment (satyabollineni) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:675]: This was the failure 

`Datalake creation failed. Data Lake creation failed on 'satya-dl4' cluster. Reason: Stack status: CREATE_FAILED, cluster status: CREATE_FAILED, reason: Failed: Orchestrator component went failed in 7 min(s), message: Salt execution went wrong:: ["Rendering SLS 'base:cloudera.agent' failed: Jinja variable 'dict object' has no attribute ''"]`

Comment (lacikaaa) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:675]: this looks more like:
- missing pillar info from the host
- faulty salt state

and not related to changing this to static
